### PR TITLE
Access EEPROM through debug interface

### PR DESCRIPTION
**Pull request by asmodai27**
# Description

Add a debug command to directly read/write on the I²C addressable EEPROM. This should allow troubleshooting all of our future usages of said EEPROM (storage of calibration parameters, on-device test results, device history and anything we can think of).

Tested on my own pizza build. Testing requires the nucleo board and interim daughterboard.

# General checklist:

- [x] I have performed a self-review, including - looked through the `Files changed` tab, browsed repository in my branch
- [x] I have made corresponding changes to the documentation - to reflect changes in code, electrical or mechanical design
- [x] All new documentation or graphics follow the [documentation style guide](https://github.com/RespiraWorks/Ventilator/wiki/Documentation-style-guide)
- [x] All new content is linked, easily discoverable, does not require too many clicks
- [x] I have reviewed any other relevant parts of our [contributor wiki](https://github.com/RespiraWorks/Ventilator/wiki) to ensure that this PR conforms to the standards
- [x] I have given the PR a descriptive name (prefixed with **WIP** if it is not yet ready for review)
- [x] I have tagged relevant reviewers

## Code checklist:

- [x] All new code follows our [code style](https://github.com/RespiraWorks/Ventilator/wiki/Code-style)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have described tests I used to verify my changes, instructions for reproducing them and any relevant configuration details.
- [x] I ran the unit test suite and verified that all tests pass - new and old, locally and on CI
- [x] I performed a clean build and verified that there were no warnings
- [x] I ran our static analysis tools and verified there were no warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: software/controller/lib/debug/debug_types.h [hunk 1]
```diff
@@ -60,6 +63,8 @@ struct Context {
                                       // can use (enforces response limits)
   uint32_t response_length; // (actual) length of the response once the
                             // command is processed
+  bool *processed; // pointer to a boolean that informs the interface handler
+                   // that a command is processed (some commands take time)
```
Comment (martukas) [inline software/controller/lib/debug/debug_types.h:67]: is being processed?
i.e. in progress?

File: software/controller/lib/debug/eeprom_cmd.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+#include "commands.h"
+
+namespace Debug::Command {
+
+ErrorCode EepromHandler::Process(Context *context) {
+
+  // We have at least subcommand and address (3 bytes)
```
Comment (martukas) [inline software/controller/lib/debug/eeprom_cmd.cpp:21]: Could we have an explicit definition of this subcommand+address data structure somewhere? Doesn't have to be a struct, but just to document it so we have a mental picture before reading all of the code.

File: software/controller/lib/debug/debug_types.h [hunk 3]
```diff
@@ -37,17 +37,20 @@ enum class ErrorCode : uint8_t {
   kInternalError = 0x05,   // Some type of internal error (aka bug)
   kUnknownVariable = 0x06, // The requested variable ID is invalid
   kInvalidData = 0x07,     // data is out of range
+  kWait = 0x08,            // awaiting peripheral response
```
Comment (martukas) [inline software/controller/lib/debug/debug_types.h:40]: Is waiting a really an error? It seems to be a non-error state, which is also defined above as a state?

File: software/controller/lib/debug/eeprom_cmd.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+      return ErrorCode::kNoMemory;
+
+    if (eeprom_->ReadBytes(address, length, context->response,
+                           context->processed)) {
+      // only set response_length if the read has been successfully sent
+      context->response_length = length;
+    }
+    return ErrorCode::kWait;
+
+  } else if (context->request[0] == 1) {
```
Comment (martukas) [inline software/controller/lib/debug/eeprom_cmd.cpp:44]: so, if I understand this correctly `request[0]` is the subcommand? And `0` is read while `1` is write? This meaning is a bit unclear.
Also, could use some consts for these subcommand definitions?

File: software/controller/lib/debug/eeprom_cmd.cpp [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    // at least one byte of data is given after the address
+    if (context->request_length < 4)
+      return ErrorCode::kMissingData;
+
+    // length of data to be written is the length of the request minus
+    // the subcommand and address bytes
+    uint16_t length = static_cast<uint16_t>(context->request_length - 3);
+
+    if (length > kMaxWriteLength)
```
Comment (martukas) [inline software/controller/lib/debug/eeprom_cmd.cpp:53]: Tracking down where `kMaxWriteLength` is defined made me realize the definitions for these handler subclasses are all in one header, and that two of them are even implemented there. Any chance we could move those implementations out into cpps and thus get them tested as well? Also, not sure if I am just being dogmatic about this, but should the definitions all be in one header?

File: software/controller/lib/debug/init.cpp [hunk 6]
```diff
@@ -59,13 +59,22 @@ static FnDebugVar vartrace_var4(
...
+I2Ceeprom eeprom = I2Ceeprom(0x50, 64, 32768, &i2c1);
+
 // Create a handler for each of the known commands that the Debug Handler can
 // link to.  This is a bit tedious but I can't find a simpler way.
 Debug::Command::ModeHandler mode_command;
 Debug::Command::PeekHandler peek_command;
 Debug::Command::PokeHandler poke_command;
 Debug::Command::VarHandler var_command;
 Debug::Command::TraceHandler trace_command(&trace);
+#ifndef TEST_MODE
```
Comment (martukas) [inline software/controller/lib/debug/init.cpp:72]: I thought we would not be doing this with ifdefs anymore?

File: software/controller/test/debug/interface_test.cpp [hunk 7]
```diff
@@ -148,6 +149,31 @@ TEST(Interface, GetVar) {
   }
 }
 
+TEST(Interface, AwaitingResponseState) {
+  Interface serial;
+  // EEPROM read command needs time to be processed
+  std::vector<uint8_t> req = {
+      static_cast<uint8_t>(Command::Code::kEepromAccess),
+      0, // Read subcommand
+      0, // address LSB
```
Comment (martukas) [inline software/controller/test/debug/interface_test.cpp:158]: what is LSB and MSB?

**Review by martukas — CHANGES_REQUESTED** [review]
Not sure if I should be strict with requiring those changes, since I am not certain what exactly those changes must be, but there are a few spots worth discussing?

**Review by asmodai27 — COMMENTED** [review]
Will take those into account except the separation of commands.h file, on which I want to discuss the best solution with you first.

**Pull request merged**